File: android/src/main/java/com/couchbase/lite/RNAuthenticator.kt
package com.couchbase.lite

import arrow.core.Either
import arrow.core.extensions.either.applicative.applicative
import arrow.core.fix
import arrow.core.flatMap
import com.facebook.react.bridge.ReadableMap

object RNAuthenticator {
  fun decode(obj: SafeReadableMap): Either<String, Authenticator> =
    obj.getString("tag")
      .flatMap { tag -> when (tag) {
        "BasicAuthenticator" -> Either.applicative<String>().tupled(
          obj.getString("username"),
          obj.getString("password")
        ).fix().map { BasicAuthenticator(it.a, it.b) }
        "SessionAuthenticator" -> obj.getString("session")
          .map { SessionAuthenticator(it) }
        else -> Either.left("$tag is not Authenticator tag")
      } }
}

File: ios/RNAuthenticator.swift
import Bow
import CouchbaseLiteSwift

class RNAuthenticator {
  static func decode(_ obj: SafeDictionary) -> Either<String, Authenticator> {
    return obj.getString("tag")
      .flatMap({ tag in
        switch tag {
        case "BasicAuthenticator":
          return Either<String, Authenticator>.map(
            obj.getString("username"),
            obj.getString("password"),
            { BasicAuthenticator(username: $0, password: $1) }
          )^
        case "SessionAuthenticator":
          return obj.getString("session")
            .map({ SessionAuthenticator(sessionID: $0) })^
        default:
          return Either.left("\(tag) is not a Authenticator tag")
        }
      })^
  }
}
